text/0000-import-trait-associated-functions.md

```diff
@@ -0,0 +1,235 @@
+
+# Summary
+[summary]: #summary
+
+Allow importing associated functions from traits and then using them like regular functions. 
+
+# Motivation
+[motivation]: #motivation
+
+There has for a long time been a desire to shorten the duplication needed to access certain associated functions, such as `Default::default`. Codebases like [Bevy](https://github.com/bevyengine/bevy/blob/7c7d1e8a6442a4258896b6c605beb1bf50399396/crates/bevy_utils/src/default.rs#L27) provide wrapper functions to shorten this call, and a previous, now-rejected, [RFC](https://github.com/rust-lang/rust/pull/73001) aimed to provide this function as part of the standard library. This RFC was rejected with a note that there is a desire for a more general capability to import any trait associated functions.
+
+Additionally, if you pull in a crate like [num_traits](https://docs.rs/num-traits/latest/num_traits/), then this feature will allow access to numeric functions such as `sin` using the `sin(x)` syntax that is more common in mathematics. More generally, it will make calls to trait associated functions shorter without having to write a wrapper function.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Importing an associated functions from a trait is the same as importing a function from any module:
+```rust
+use Default::default;
+```
+
+Once you have done this, the function is made available in your current scope just like any other regular function.
+
+```rust
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: default()
+        }
+    }
+}
+```
+
+You can also use this with trait methods (i.e. associated functions that take a `self` argument):
+
+```rust
+use num_traits::float::Float::{sin, cos}
+
+fn eulers_formula(theta: f64) -> (f64, f64) {
+    (cos(theta), sin(theta))
+}
+```
+
+Importing an associated function from a trait does not import the trait. If you want to call `self` methods on a trait or `impl` it, then you can import the trait and its associated functions in a single import statement:
+
+```rust
+mod a {
+    trait A {
+        fn new() -> Self;
+        fn do_something(&self);
+    }
+}
+
+mod b {
+    use super::a::A::{self, new}
+
+    struct B();
+
+    impl A for B {
+        fn new() -> Self {
+            B()
+        }
+
+        fn do_something(&self) {
+        }
+    }
+
+    fn f() {
+        let b: B = new();
+        b.do_something();
+    }
+}
+```
+
+Associated functions can also be renamed when they are imported using the usual `as` syntax:
+```rust
+use Default::default as gimme
+
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: gimme()
+        }
+    }
+}
+```
+
+You cannot import a parent trait associated function from a sub-trait:
+
+```rust
+use num_traits::float::Float::zero; // Error: try `use num_traits::identities::Zero::zero` instead.
```

Comment:
This seems like a notable limitation. We've talked about adding a compatibility mechanism that allows *implementing* supertrait methods via the subtrait, such that it would then be possible to extract a method from a trait into a supertrait without breaking compatibility with any user of the trait. This limitation would prevent that.

So, at the very least, *if* we add such a mechanism we'll have to fix this limitation at that time as well. Could you add a note to future work to that effect? "If we add a compatibility mechanism to implement a supertrait method when implementing its subtrait, without having to separately implement the supertrait (such that a new supertrait can be extracted from a trait without breaking compatibility), we would also need to lift the limitation on using a supertrait method via a subtrait."